### ISSUE TITLE
The inclusion of the Portuguese States (Distritos) in 5.2.0 can potentially break 4000+ websites

### ISSUE THREAD

**Issue by webdados** [woocommerce/woocommerce#29667]
A user opened issue #29206 requesting the addition of the Portuguese states (known as Distritos) on the WooCommerce core, with the argument it needs to be used for taxes.

Never a Portuguese store owner had this problem in years and years of WooCommerce existence because this problem does not exist. There are in fact different taxes in the mainland and the islands but the Portuguese tax law states that the VAT is charged base on the shop address, not the customer's address. In sum: Portuguese states are never used for tax calculations.

For full disclosure, I'm the author of a well-established [plugin that adds the Portuguese states (Distritos) to WooCommerce](https://wordpress.org/plugins/portugal-states-distritos-for-woocommerce/) [for 7 years now](https://plugins.trac.wordpress.org/log/portugal-states-distritos-for-woocommerce/).
This plugin was launched with the objective of using Districts to create shipping zones for shipping calculation proposes (not taxes).

Although I always use the most recent development versions of WooCommerce on my dev environment, I was not aware of this change because this skipped the beta and RC versions and went directly to the 5.2.0 stable version.
The user that submitted the issue is well intended, but he's not Portuguese, doesn't know the reality of the Portuguese tax laws, seems to not be aware of the Portuguese WooCommerce ecosystem, and this was his first opened issue not only on WooComerce but on Github altogether, and yet this was released faster than anything I ever seen be released on the WooCommerce core.

Besides, of course, seeing my (free) work of 7 years being rendered obsolete, there are things that concern me more about this addition to the core:

- AFAIK, the reason WooCommerce didn't include the Portuguese districts (or other countries states) it's because they're not textually part of the written postal address https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-wc-countries.php#L540, or are not used for tax calculations, so IMHO I think "Portuguese states" is plugin territory and not core.

- Because a plugin already exists, there are 4000+ Portuguese shops that already have the states code in their thousands of orders, the WooCommerce shop country/region setting, shipping zones settings, customers saved addresses (and I can't even imagine where else on the database). The same array keys were not used: https://plugins.trac.wordpress.org/browser/portugal-states-distritos-for-woocommerce/trunk/woocommerce-portugal-states.php#L69 vs https://github.com/woocommerce/woocommerce/issues/29206#L1299 so this means if a user has been using this plugin for years and then realizes WooCommerce is now shipping with the Portuguese districts, he will probably de-activate the plugin and a LOT of stuff will break on his store.

In conclusion:
My suggestion is that #29586 is quickly removed from the core before breaking thousands of stores, and also because taking into account the reasons WooCommerce adds (or doesn't add) countries states (IMHO) are not met for Portugal.
If the WooCommerce team decides to keep this on the core, at least mimic the array keys we used on our plugin and avoid breaking a lot of stuff for thousands of Portuguese store owners and their clients.

**Comment by pedro-mendonca** [comment]
This is a real concern, as currently most of the Portuguese WooCommerce shops uses the States data from [Portugal States (Distritos) for WooCommerce](https://wordpress.org/plugins/portugal-states-distritos-for-woocommerce/).

This really **should** be reverted (https://github.com/woocommerce/woocommerce/commit/5b274fda874607e0334f54f75d768b727effe780) and reconsidered against the currently existent solution without any major breaking changes.

Thanks @webdados bringing this up.

**Comment by gumbapt** [comment]
The issue #29206 specifically states that he needs that to use in district taxes in... Portugal? Like Webdados explained, we don't use that in Portugal so I wonder what was the original requester was talking about. 

At very least, it should use the same keys if not reverted.

**Comment by MarcosCoelho234** [comment]
I have three Portuguese stores with Woocommerce and I cannot upgrade to 5.2.0 for the reason mentioned by @webdados.

The latest update will have a **negative impact on all Portuguese stores**.

**Comment by vedanshujain** [comment]
@webdados looks like this was an unintentional change that was slated for 5.3. We will be soon releasing a fix release 5.2.1 which will not have this change.

For now, we will also revert the change for 5.3, but we do need to be consistent with CLDR notation.

**Comment by webdados** [comment]
> I have three Portuguese stores with Woocommerce and I cannot upgrade to 5.2.0 for the reason mentioned by @webdados.
> 
> The latest update will have a **negative impact on all Portuguese stores**.

You can update, but if you deactivate the Distritos plugin all your orders, addresses and settings related to the Portuguese districts will lose that information.

**Comment by webdados** [comment]
> @webdados looks like this was an unintentional change that was slated for 5.3. We will be soon releasing a fix release 5.2.1 which will not have this change.
> 
> For now, we will also revert the change for 5.3, but we do need to be consistent with CLDR notation.

Hi @vedanshujain thanks for your reply.

I understand the need to be consistent with the CLDR notation, but back when I developed the plugin (7 years ago), no country in WooCommerce had its states defined with that notation, as we can easily see in the current codebase.

So to be clear:
- 5.2.1 will remove the "Portuguese states"
- 5.3.0 will reinstate this change
Is that it?

**Comment by vedanshujain** [comment]
@webdados 
The current plan is:
1. 5.2.1 will remove the "Portuguese states"
2. 5.3.0 will "revert" this change, so it won't have it either

but we are also discussing internally if there is a good way to go about these changes and (most likely) discuss that publicly as well on the dev blog.

**Comment by webdados** [comment]
Still not sure about the difference of "removing" on 5.2.1 and "reverting" on 5.3.0 :-D
I guess I'm missing something here.

Anyway, first of all, thanks for being quick on addressing this potential problem for 4000+ shops in Portugal.

Regarding the future, and because the discussion is public, taking into consideration that 1) the district is not part of the Portuguese postal address format and 2) we do not need to make tax configurations based on the district, my opinion is that the Portuguese districts should not ship with the WooCommerce core (as it happens with several other countries that meet the same criteria).
Yes, I did develop a plugin 7 years ago to add this to WooCommerce, but for the sake of improving the shop owner experience while creating shipping zones based on the district (mainly to separate the islands Azores and Madeira from the mainland), but this is a nice-to-have, a perfect example of "plugin territory".

**Comment by Buzut** [comment]
There are different things to consider here. I am not 100% sure of the rule guiding the inclusion or not of some countries' states/regions/provinces/districts or not. 

For reference, I think there are some inconsistencies. For instance: all Spanish regions are listed but Portugal as well as France have none. Nevertheless, unlike Portugal, France has its overseas territories listed as standalone countries (French Polynesia, French Guiana…).

Now there are two reasons why one would want to target a region: tax computation or shipping calculation. 

While the same tax applies everywhere in Portugal, shipping to Madeira isn't nearly the same as shipping to Porto, hence, I think it would be a great addition to the core – and extremely logic – to at least integrate the two Autonomous Regions of Portugal in one shape or another.

As a side note, there is also the [WooCommerce Locations Pack plugin]() that adds locations (Portuguese districts, French departments, UK counties…).

Also note that these inconsistencies automatically make Spanish and Portuguese territories part of Europe as a continent whereas they technically aren't located in Europe. On the other hand, French territories being separate countries won't automatically fall under an Europe shipping zone.

**Comment by webdados** [comment]
@Buzut read the discussion above. We don’t want to break 5000+ stores, do we?
Shipping calculation is already possible using post codes interval or using the linked plugin.
The district is not part of the written Portuguese format neither there’s tax calculations that need to be done depending on the customer location, so if it’s only for shipping, you have two solutions already.

**Comment by Buzut** [comment]
Actually, I had already read the whole discussion. Adopting the same way as the one used for French overseas territories (ie. listing them as different countries) wouldn't break the stores using your plugin, would it? I am talking about adding the two autonomous regions, not adding all "Distritos".

It would be exactly the same logic used for France. There is no way to specify Départements or Régions, but overseas are something different, so they're listed as standalone countries.

Sure plugins offer a solution, that's not the question here. The question is whether it is sufficiently important to deserve to be integrated into the core. **The answer is yes**. If you can target Spanish special territories and French overseas specifically, I don't see why the core shouldn't allow the same behaviour for Portuguese ones.

**Comment by webdados** [comment]
The French example is not the same at all. You're talking about different continents.
The Portuguese islands are on the same territory as the mainland, as the Spanish ones are (also not separated in the list of countries, as they shouldn't).

For example, French Polynesia is considered a country: https://en.wikipedia.org/wiki/French_Polynesia
French Guiana has its own ISO code: https://dgn.isolutions.iso.org/obp/ui#iso:code:3166:GF
I don't know enough (neither do I want to) about the French political and territory organization, but if a territory is not considered a country it shouldn't be seen as one on WooCommerce. Just because we might have a bad example (which should probably be revised) it doesn't mean we should start messing up the countries list.

Madeira and Açores are not considered countries nor do have ISO codes.

What you're proposing is to subvert the concept of a country completely by separating islands from their countries, and that's a path I hope no one with merge privileges on WooCommerce ever considers.

**Comment by Buzut** [comment]
> Just because we might have a bad example (which should probably be revised) it doesn't mean we should start messing up the countries list.

Will never be revised for the exact same reason you mentioned regarding your plugin: breaking changes.

Polynesia is the only exemple (of overseas country) and the status is recent. Most other French overseas are Region/Departments.

I agree with you though that we shouldn't have entries in the countries list that aren't technically countries. What I regret is the lack of proper addressing mechanism regarding shipping.

Limiting a zone to certain zip codes is just unmanageable and barely possible if you don't have a deep knowledge of a particular country's codes.

What I think would be ideal is to integrate every country's region (like Portugal's distritos, UK's counties…) and have a flag indicating if they should appear in the address part or not.

Again, this would probably break backwards compatibility and not be easy to integrate at WC scale. At least for the near future, there's the [WooCommerce Locations Pack](https://wordpress.org/plugins/woo-locations/) plugin or `woocommerce_states` filter.
